androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LocationProviderGmsCore.java

```diff
@@ -0,0 +1,190 @@
+// Copyright 2020 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.androidbrowserhelper.trusted;
+
+import android.content.Context;
+import android.location.Location;
+import android.os.Bundle;
+import android.os.Looper;
+import android.os.RemoteException;
+import android.util.Log;
+
+import com.google.android.gms.common.ConnectionResult;
+import com.google.android.gms.common.GoogleApiAvailability;
+import com.google.android.gms.common.api.GoogleApiClient;
+import com.google.android.gms.common.api.GoogleApiClient.ConnectionCallbacks;
+import com.google.android.gms.common.api.GoogleApiClient.OnConnectionFailedListener;
+import com.google.android.gms.location.FusedLocationProviderApi;
+import com.google.android.gms.location.LocationListener;
+import com.google.android.gms.location.LocationRequest;
+import com.google.android.gms.location.LocationServices;
+
+import androidx.browser.trusted.TrustedWebActivityCallbackRemote;
+
+/**
+ * This is a LocationProvider using Google Play Services.
+ * <p>
+ * https://developers.google.com/android/reference/com/google/android/gms/location/package-summary
+ */
+public class LocationProviderGmsCore
+        implements ConnectionCallbacks, OnConnectionFailedListener, LocationListener {
+    private static final String TAG = "TWA_LocationProviderGms";
+
+    private static final String EXTRA_NEW_LOCATION_AVAILABLE_CALLBACK = "onNewLocationAvailable";
+    private static final String EXTRA_NEW_ERROR_AVAILABLE_CALLBACK = "onNewErrorAvailable";
+
+    // Values for the LocationRequest's setInterval for normal and high accuracy, respectively.
+    private static final long UPDATE_INTERVAL_MS = 1000;
+    private static final long UPDATE_INTERVAL_FAST_MS = 500;
+
+    private static LocationProviderGmsCore sProvider;
+    private final GoogleApiClient mGoogleApiClient;
+    private FusedLocationProviderApi mLocationProviderApi = LocationServices.FusedLocationApi;
+
+    private TrustedWebActivityCallbackRemote mCallback;
+
+    private boolean mEnablehighAccuracy;
+    private LocationRequest mLocationRequest;
+
+    LocationProviderGmsCore(Context context) {
```

Comment:
What happens if a web page requests location and doesn't "stop" it? And what happens if A page requests location multiple times? I'm wondering whether we need to be wary of any state left around.